File: content/dota_addons/ziv/panorama/scripts/custom_game/network.js
"use strict";

var Util = GameUI.CustomUIConfig().Util;

var authKey = null;
var url = null;

function SendRequest( requestParams, successCallback )
{
    if (!authKey || ! url)
    {
        $.Msg('Auth key or URL is empty!');
        return;
    }

    requestParams.AuthKey = authKey;

    $.AsyncWebRequest(url,
        {
            type: 'POST',
            data: { 
                CommandParams: JSON.stringify(requestParams) 
            },
            success: successCallback,
            error: (function(data){
                $.Msg(data);
            })
        });
}

function SetAuthParams( args )
{
    authKey = args.AuthKey;
    url = args.URL;
}

function LoadCharacters( callback ) {
    var requestParams = {
        Command: "LoadCharacters",
        SteamID: Util.GetSteamID32()
    };

    GameUI.CustomUIConfig().SendRequest( requestParams, callback );    
}

function DeleteCharacter( ID, callback ) {
    var requestParams = {
        Command: "DeleteCharacter",
        ID: ID
    };

    GameUI.CustomUIConfig().SendRequest( requestParams, callback );    
}

(function() {
    GameUI.CustomUIConfig().SendRequest = SendRequest;
    GameUI.CustomUIConfig().LoadCharacters = LoadCharacters;
    GameUI.CustomUIConfig().DeleteCharacter = DeleteCharacter;
    GameEvents.Subscribe( "su_auth_params", SetAuthParams );
})();

File: game/dota_addons/ziv/scripts/vscripts/libraries/network.lua
SU = {}

SU.STAT_SETTINGS = LoadKeyValues('scripts/kv/StatUploaderSettings.kv')
SU.is_stat_server_up = true

-- Request function
function SU:SendRequest( request_params, success_callback )
  if not SU.is_stat_server_up then
    return
  end
  
  -- Adding auth key
  request_params.AuthKey = SU.STAT_SETTINGS.AuthKey
  -- DeepPrintTable(request_params)

  -- Create the request
  local request = CreateHTTPRequest('POST', SU.STAT_SETTINGS.Host)
  request:SetHTTPRequestGetOrPostParameter('CommandParams', json.encode(request_params))

  -- Send the request
  request:Send(function(res)
    if res.StatusCode ~= 200 or not res.Body then
        -- Bad request
        SU.is_stat_server_up = false
        
        print("Request error. See info below: ")
        DeepPrintTable(res)
        return
    end

    -- Try to decode the result
    local obj, pos, err = json.decode(res.Body, 1, nil)
    
    -- if not a JSON send full body
    if obj == nil then
      obj = res.Body
    end
    
    -- Feed the result into our callback
    success_callback(obj)
  end)
end

function SU:SendAuthInfo()
  CustomGameEventManager:Send_ServerToAllClients("su_auth_params", { URL = SU.STAT_SETTINGS.Host, AuthKey = SU.STAT_SETTINGS.AuthKey } )
end

-------------------------------------------------------------------------------
-- Testing functions
-------------------------------------------------------------------------------

function SU:Test()
  local request_params = {
    Command = "Testing"
  }
    
  SU:SendRequest( request_params, function(obj)
      print("Testing result: ", obj)
  end)
end

-------------------------------------------------------------------------------
-- Utilities
-------------------------------------------------------------------------------
table.filter = function(t, filter_iter)
  local out = {}

  for k, v in pairs(t) do
    if filter_iter(k, v, t) then out[k] = v end
  end

  return out
end